### PR TITLE
Adding validation in autoscaler

### PR DESCRIPTION
**Pull request by truelegion47**
## Why are these changes needed?

[#23166](https://github.com/ray-project/ray/issues/23166) needed type validation in autoscaler for bundles.

## Related issue number

ISSUE-23166

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_autoscaler.py [hunk 1]
```diff
@@ -3561,6 +3561,17 @@ def __init__(self, *args, **kwargs):
                     monitor.run()
                 mock_publish.assert_called_once()
 
+    def testAutoscalerResourceValidation(self):
+        from ray.autoscaler.sdk import request_resources
+
+        self.assertRaises(TypeError, request_resources(bundles={"CPU": 1}))
+        self.assertRaises(TypeError, request_resources(bundles=[{"GPU": 1}]))
```
Comment (truelegion47) [inline python/ray/tests/test_autoscaler.py:3568]: @DmitriGekhtman Is `[{"GPU": 1}]` a valid bundle value or we always would require "CPU" to be present ?

Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:3568]: `[{"GPU": 1}]` is valid.

File: python/ray/autoscaler/sdk/sdk.py [hunk 2]
```diff
@@ -240,6 +240,24 @@ def request_resources(
...
     """
+
+    if num_cpus is not None and not isinstance(num_cpus, int):
+        raise TypeError("num_cpu must be either None or an int.")
+    if bundles is not None:
+        if isinstance(bundles, List):
+            for bundle in bundles:
+                if isinstance(bundle, Dict):
+                    if "CPU" not in bundle or "GPU" not in bundle:
+                        raise TypeError("CPU or GPU key should be present in bundle.")
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/sdk/sdk.py:251]: Actually, any string key is valid. So we should just validate that the keys are strings.

A node can be labelled with custom resource capacities. like "CUSTOM": 10.
Then you can for example
schedule 10 tasks requesting {"CUSTOM": 1} on such a node, or
trigger upscaling of 2 such nodes by requesting two {"CUSTOM": 10} bundles
`ray.autoscaler.sdk.request_resources(bundles=[{"CUSTOM":10}]*2)`

File: python/ray/experimental/state/api.py [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
 import warnings
-import threading
-import logging
-from dataclasses import fields
-from typing import Dict, Generator, List, Optional, Tuple, Union, Any
 from contextlib import contextmanager
+from dataclasses import fields
+from typing import Any, Dict, Generator, List, Optional, Tuple, Union
 
 import requests
```
Comment (truelegion47) [inline python/ray/experimental/state/api.py:9]: Fix checkstyle.

**Comment by truelegion47** [comment]
@DmitriGekhtman Can you please take a look when you get a chance ?

File: python/ray/autoscaler/sdk/sdk.py [hunk 4]
```diff
@@ -240,6 +240,23 @@ def request_resources(
         >>> request_resources( # doctest: +SKIP
         ...     bundles=[{"CPU": 1}, {"CPU": 1}, {"CPU": 1}])
     """
+
+    if num_cpus is not None and not isinstance(num_cpus, int):
+        raise TypeError("num_cpu must be either None or an int.")
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/sdk/sdk.py:245]: nit
```suggestion
        raise TypeError("num_cpus must be either None or an int.")
```

File: python/ray/autoscaler/sdk/sdk.py [hunk 5]
```diff
@@ -240,6 +240,23 @@ def request_resources(
...
+        if isinstance(bundles, List):
+            for bundle in bundles:
+                if isinstance(bundle, Dict):
+                    for key in bundle.keys():
+                        if not (isinstance(key, str) and isinstance(bundle[key], int)):
+                            raise TypeError(
+                                "bundle key must be string and value must be int."
+                            )
+                else:
+                    raise TypeError("bundle must be of type Dict.")
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/sdk/sdk.py:256]: nit
```suggestion
                    raise TypeError("each bundle must be of type Dict.")
```

File: python/ray/autoscaler/sdk/sdk.py [hunk 6]
```diff
@@ -240,6 +240,23 @@ def request_resources(
...
+    if num_cpus is not None and not isinstance(num_cpus, int):
+        raise TypeError("num_cpu must be either None or an int.")
+    if bundles is not None:
+        if isinstance(bundles, List):
+            for bundle in bundles:
+                if isinstance(bundle, Dict):
+                    for key in bundle.keys():
+                        if not (isinstance(key, str) and isinstance(bundle[key], int)):
+                            raise TypeError(
+                                "bundle key must be string and value must be int."
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/sdk/sdk.py:253]: nit
```suggestion
                                "each bundle key must be string and value must be int."
```

**Review by DmitriGekhtman — APPROVED** [review]
Left some final nits to address before merging.

Thank you for the fix!

**Comment by DmitriGekhtman** [comment]
After addressing the final nits, would you mind also fixing the DCO check? The Details link above will give you instructions on how to rebase to fix it.

**Pull request closed without merging**